### PR TITLE
AD Kernel{Value,Grad}

### LINKED ISSUE #12658: Create ADKernelValue and ADKernelGrad
## Rationale

The `KernelValue` and `KernelGrad` classes provide a `precomputeQpResidual` method to override which computes the part of the residual that is independent of the test function index `_i`. That result is then multiplied by all test functions (or their gradients respectively). This approach allows a loop reordering causing the residual to be re-computed significantly fewer times. This optimization would be even more significant in the AD system when in the JACOBIAN compute stage.

## Description

Implement intermediate base class templated that replicate the loop reordering optimization of `KernelValue` and `KernelGrad`.

## Impact

Faster kernel evaluation for residuals which allow the test function or the test function gradient to be factored out.

Tag @lindsayad

### PR DESCRIPTION
**Pull request by dschwen**
My stab at creating automatic differentiation versions of KernelValue and KernelGrad.
Builds on stuff in <s>#12650</s> #12657 (so it'd be great if that would get reviewed, too)
 
Closes #12658

### CODE REVIEW

**Comment by lindsayad** [comment]
> Builds on stuff in #12650 (so it'd be great if that would get reviewed, too)

You want me to review the issue? 😄

**Comment by dschwen** [comment]
```
SQA ERROR: Failed to locate the design page '/syntax/Kernels/index.md'
```

https://github.com/idaholab/moose/blob/next/framework/doc/content/syntax/Kernels/index.md

Whiskey. Tango. Foxtrot.

**Comment by dschwen** [comment]
What is going on here?!!!

```
outputs/format.dump_test: ################################################################################
outputs/format.dump_test: 
outputs/format.dump_test: Unable to match the following pattern against the program's output:
outputs/format.dump_test: 
outputs/format.dump_test: block\s*=\s*ANY_BLOCK_ID
outputs/format.dump_test: 
outputs/format.dump_test: ################################################################################
outputs/format.dump_test: Tester failed, reason: EXPECTED OUTPUT MISSING
outputs/format.dump_test: 
    FAIL outputs/format.dump_test FAILED (EXPECTED OUTPUT MISSING)
```

This seems unrelated.

**Comment by dschwen** [comment]
And it makes no sense!!!!

```
[12][schwd][~/Programs/moose/test/tests/outputs/format]> ../../../moose_test-opt --dump|grep ANY_BLOCK
          block              = ANY_BLOCK_ID        # block ID or name where the object works
        block                                = ANY_BLOCK_ID          # The list of block ids (SubdomainID)
[12][schwd][~/Programs/moose/test/tests/outputs/format]> grep ANY tests
    expect_out = 'block\s*=\s*ANY_BLOCK_ID'
```

Why is the regexp not matching?!!

**Comment by lindsayad** [comment]
> Why is the regexp not matching?!!

I've run into this before for this test. I think it was something about the test harness trimming the output. @permcody @milljm ?

**Comment by lindsayad** [comment]
Looks like #12171 is relevant here...

**Comment by lindsayad** [comment]
Try adding `max_buffer_size = -1` for that test?

**Comment by permcody** [comment]
I'd prefer you don't remove the max buffer size. If the pattern you are looking for is getting trimmed, your test is probably too long. You can certainly try that as a debugging tool though.

**Comment by dschwen** [comment]
I'm not looking for anything. This is an unrelated pre-existing test!

**Comment by permcody** [comment]
Here's how the trimming works:
If your output buffer is more than 10K, we trim out the middle. We keep the first 2/3rds and the last 1/3rd of the total output. It's pretty unlikely that a pattern that you are looking for is right in the middle (trimmed region).

**Comment by dschwen** [comment]
☘️

**Comment by permcody** [comment]
We don't cover pre-existing tests 😆, call your developers and complain.

File: framework/include/materials/ADMaterial.h [hunk 1]
```diff
@@ -17,16 +17,22 @@
   using ADMaterial<compute_stage>::_qp;                                                            \
   using ADMaterial<compute_stage>::_ad_grad_zero;                                                  \
   using ADMaterial<compute_stage>::_ad_zero;                                                       \
-  using ADMaterial<compute_stage>::coupledComponents
+  using ADMaterial<compute_stage>::_qrule;                                                         \
+  using ADMaterial<compute_stage>::_JxW;                                                           \
+  using ADMaterial<compute_stage>::_coord;                                                         \
+  using ADMaterial<compute_stage>::_assembly;                                                      \
+  using ADMaterial<compute_stage>::_mesh;                                                          \
```
Comment (tophmatthews) [inline framework/include/materials/ADMaterial.h]: These will conflict with #12649...it's a race!

Comment (dschwen) [reply framework/include/materials/ADMaterial.h]: As you can see from the constantly failing tests I'm merely limping along...

File: framework/doc/content/source/kernels/Kernel.md [hunk 2]
```diff
@@ -0,0 +1,14 @@
+# Kernel
+
+Kernel is the base class for object that calculate residual vector contributions
+of non-linear scalar field variables.
+
+Derived from it are two sets of optimized base classes that - through loop
+reordering - reduce the number of necessary residual evaluations and enable the
+compiler to perform code vectorization. These base classes have restrictions on
+the mathematical for of the residuals they can be applied to.
```
Comment (lindsayad) [inline framework/doc/content/source/kernels/Kernel.md]: `for` -> `form`

File: framework/src/kernels/ADKernelGrad.C [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+template <>
+void
+ADKernelGrad<RESIDUAL>::computeJacobian()
+{
+}
+
+template <ComputeStage compute_stage>
+void
+ADKernelGrad<compute_stage>::computeOffDiagJacobian(MooseVariableFEBase & jvar)
```
Comment (lindsayad) [inline framework/src/kernels/ADKernelGrad.C]: This is outside the scope of this PR, but another optimization that will need to be done in the future is to actually have the variable loop actually be an inner loop that's called after calculating the residuals and then we're just doing the fill. I have this code in my `laser-welding` branch which I will incorporate into upstream soon:
```c++
template <typename T, ComputeStage compute_stage>
void
ADKernelTempl<T, compute_stage>::computeADOffDiagJacobian()
{
  std::vector<ADReal> residuals(_test.size(), 0);

  beforeQpLoop();
  for (_qp = 0; _qp < _qrule->n_points(); _qp++)
  {
    beforeTestLoop();
    for (_i = 0; _i < _test.size(); _i++)
      residuals[_i] += _ad_JxW[_qp] * _coord[_qp] * computeQpResidual();
  }

  std::vector<std::pair<MooseVariableFEBase *, MooseVariableFEBase *>> & ce =
      _assembly.couplingEntries();
  for (const auto & it : ce)
  {
    MooseVariableFEBase & ivariable = *(it.first);
    MooseVariableFEBase & jvariable = *(it.second);

    unsigned int ivar = ivariable.number();
    unsigned int jvar = jvariable.number();

    if (ivar != _var.number())
      continue;

    size_t ad_offset = jvar * _sys.getMaxVarNDofsPerElem();

    prepareMatrixTag(_assembly, ivar, jvar);

    if (_local_ke.m() != _test.size() || _local_ke.n() != jvariable.phiSize())
      continue;

    for (_i = 0; _i < _test.size(); _i++)
      for (_j = 0; _j < jvariable.phiSize(); _j++)
        _local_ke(_i, _j) += residuals[_i].derivatives()[ad_offset + _j];

    accumulateTaggedLocalMatrix();
  }
}
```

File: framework/include/kernels/ADKernelGrad.h [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+  virtual void computeJacobian() override;
+  virtual void computeOffDiagJacobian(MooseVariableFEBase & jvar) override;
+
+protected:
+  /**
+   * Called before forming the residual for an element
+   */
+  virtual ADGradResidual precomputeQpResidual() = 0;
+
+  virtual ADResidual computeQpResidual() final;
```
Comment (lindsayad) [inline framework/include/kernels/ADKernelGrad.h]: I wasn't even aware of this keyword. Cool!

File: framework/src/kernels/ADKernelGrad.C [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "ADKernelGrad.h"
+#include "Assembly.h"
```
Comment (lindsayad) [inline framework/src/kernels/ADKernelGrad.C]: I know this is copy-pasta but some of these includes appear unnecessary and I needed to at least one other review comment

File: framework/src/kernels/ADKernelValue.C [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "ADKernelValue.h"
+#include "Assembly.h"
```
Comment (lindsayad) [inline framework/src/kernels/ADKernelValue.C]: Same comment as above

File: test/tests/kernels/kernel_precompute/tests [hunk 7]
```diff
@@ -1,8 +1,20 @@
...
   [./test]
     type = 'Exodiff'
     input = 'kernel_precompute_test.i'
     exodiff = 'out.e'
     scale_refine = 5
+    requirement = 'MOOSE shall provide optimized kernels for residuals with the test function or its gradient factored out.'
+  [../]
+  [./adtest]
+    type = 'Exodiff'
+    input = 'kernel_precompute_test.i'
```
Comment (lindsayad) [inline test/tests/kernels/kernel_precompute/tests]: Should be `adkernel_precomute_test.i`? Also I do like to see `PetscJacobianTesters` for our AD stuff

**Review by lindsayad — CHANGES_REQUESTED** [review]
This looks great. Just a couple of small requests

**Review by lindsayad — CHANGES_REQUESTED** [review]

File: framework/src/kernels/ADKernelValue.C [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+    precalculateResidual();
+    for (_qp = 0; _qp < _qrule->n_points(); _qp++)
+    {
+      // This will also compute the derivative with respect to all dofs
+      const auto value = precomputeQpResidual() * _JxW[_qp] * _coord[_qp];
+      for (_i = 0; _i < _test.size(); _i++)
+      {
+        const auto residual = value * _test[_i][_qp];
+        for (_j = 0; _j < jvar.phiSize(); _j++)
+          _local_ke(_i, _j) += _JxW[_qp] * _coord[_qp] * residual.derivatives()[ad_offset + _j];
```
[the remark this answers is not in the document]
Comment (dschwen) [reply framework/src/kernels/ADKernelValue.C]: Ah, thanks, I didn't even dare look at it yet (didn't want to ruin my weekend). But it turned out to be pretty straight forward :-D

File: framework/doc/content/source/kernels/Kernel.md [hunk 9]
```diff
@@ -0,0 +1,14 @@
+# Kernel
+
+Kernel is the base class for object that calculate residual vector contributions
+of non-linear scalar field variables.
+
+Derived from it are two sets of optimized base classes that - through loop
+reordering - reduce the number of necessary residual evaluations and enable the
+compiler to perform code vectorization. These base classes have restrictions on
+the mathematical form of the residuals they can be applied to.
```
Comment (tophmatthews) [inline framework/doc/content/source/kernels/Kernel.md]: Is it easy enough to spell out where this doesn't work?

File: framework/doc/content/source/kernels/KernelGrad.md [hunk 10]
```diff
@@ -0,0 +1,19 @@
+# KernelGrad
+
+Derive from this base class if your residual is of the form
+
+$$
+(\dots,\nabla\psi_i)
+$$
+
+i.e. if the test function $\nabla\psi_i$ (`_grad_test[_i][_qp]`) can be factored out.
```
Comment (tophmatthews) [inline framework/doc/content/source/kernels/KernelGrad.md]: Oh I suppose you do spell it out in the individual kernels...

**Comment by dschwen** [comment]
@lindsayad, does this need a new reviewer now that you contributed?

**Review by lindsayad — APPROVED** [review]

**Comment by lindsayad** [comment]
@permcody says if github allows it we're good! 😄

**Comment by tophmatthews** [comment]
@lindsayad @dschwen so a bit on the practicality part, should I move all the kernels I use that can use this precompute over to inherit from this? It almost seems like for the exact same physics, you could have 4 different kernels (`Kernel`, `ADKernel`, `KernelValue`, `ADKernelValue`). How should these be handled? How much quicker is it really?

**Comment by lindsayad** [comment]
Yes you could have four, but I think there are clear ideas behind which one
should be chosen. If you want to automatically compute the jacobian
(sacrificing a little bit of speed compared to using an accurate hand-coded
Jacobian), then choose an `AD` object. If your residual is only going to
contain terms multiplied by _test, then use `ADKernelValue`. The speed-up
is entirely dependent on the complexity of the residual. If you're just
multiplying _test[_i][_qp] * _u[_qp], then you will save no operations (and
consequently obtain no speed benefits) by switching over from `ADKernel` to
`ADKernelValue`. But in an example case of working on Navier Stokes for
example, I experienced substantial speed-ups by doing these optimizations;
I remember factors of three or so for the entire simulation.

**Pull request merged**